sklearn/neighbors/lmnn.py

```diff
@@ -0,0 +1,1070 @@
+                                           algorithm=self.neighbors_algorithm)
+
+        # Compute the gradient part contributed by the target neighbors
+        grad_static = _compute_grad_static(X_valid, targets, self.verbose)
+
+        # Decide how to store the impostors
+        if self.imp_store == 'sparse':
+            use_sparse = True
+        elif self.imp_store == 'list':
+            use_sparse = False
+        else:
+            # auto: Use a heuristic based on the data set size
+            use_sparse = X_valid.shape[0] > 6500
+
+        # Create a dictionary of parameters to be passed to the optimizer
+        disp = self.verbose - 2 if self.verbose > 1 else -1
+        optimizer_params = {'method': 'L-BFGS-B',
+                            'fun': self._loss_grad_lbfgs,
+                            'jac': True,
+                            'args': (X_valid, y_valid, targets, grad_static,
+                                     use_sparse),
+                            'x0': transformation,
+                            'tol': self.tol,
+                            'options': dict(maxiter=self.max_iter, disp=disp),
+                            'callback': self._callback
+                            }
+
+        # Call the optimizer
+        self.n_iter_ = 0
+        opt_result = minimize(**optimizer_params)
+
+        # Reshape the solution found by the optimizer
+        self.transformation_ = opt_result.x.reshape(-1, X_valid.shape[1])
+
+        # Stop timer
+        t_train = time.time() - t_train
+        if self.verbose:
+            print('Training took {:8.2f}s.'.format(t_train))
+
+        # Optionally store information returned by the optimizer
+        if self.store_opt_result:
+            self.opt_result_ = opt_result
+
+        return self
+
+    def transform(self, X):
+        """Applies the learned transformation to the given data.
+
+        Parameters
+        ----------
+        X : array-like, shape (n_samples, n_features)
+            Data samples.
+
+        Returns
+        -------
+        X_embedded: array-like, shape (n_samples, n_features_out)
+            The data samples transformed.
+
+        Raises
+        ------
+        NotFittedError
+            If :meth:`fit` has not been called before.
+        """
+
+        return self._transform(X)
+
+    def _transform(self, X, check_input=True):
+        """Applies the learned transformation to the given data.
+
+        Parameters
+        ----------
+        X : array-like, shape (n_samples, n_features)
+            Data samples.
+
+        check_input: bool, optional (default=True)
+            Whether to validate ``X``.
+
+        Returns
+        -------
+        X_embedded: array-like, shape (n_samples, n_features_out)
+            The data samples transformed.
+
+        Raises
+        ------
+        NotFittedError
+            If :meth:`fit` has not been called before.
+        """
+
+        check_is_fitted(self, ['transformation_'])
```

Comment:
Maybe this should be moved in the public "transform" method instead of the private "_transform" method, that I believe is there for speed reasons.

I think that maybe we could even go one step further, and turn the method method to "_transform_without_checks" method, and move all the checks to the public "transform" method. I find that it makes the purpose of these methods more easy to understand (eg for the maintainer or future developer). In addition, you gain speed by not performing checks when not necessary.

Reply:
Done.